File: Python/Stack/Maximum Score From Removing Substrings.py
class Solution(object):
    def remsubstr(self, s, rem, points):
        score = 0
        st = []
        for i in s:
            if st and st[-1] == rem[0] and i == rem[1]:
                st.pop()
                score=score+points
            else:
                st.append(i)
        # Rebuild the remaining string from the stack
        remaining = ''.join(st)
        s=remaining
    
        return remaining, score
        
    def maximumGain(self, s, x, y):
        """
        :type s: str
        :type x: int
        :type y: int
        :rtype: int
        """
        points = 0
    
        if x >= y:
            s, score = self.remsubstr(s, "ab", x)
            points += score
            s, score = self.remsubstr(s, "ba", y)
            points += score
        else:
            s, score = self.remsubstr(s, "ba", y)
            points += score
            s, score = self.remsubstr(s, "ab", x)
            points += score
    
        return points

File: CPP/Stack/Maximum Score From Removing Substrings.cpp
class Solution {
public:
    int removeSubstr(string& s, string rem, int points) {
        stack<char> st;
        int score = 0;
        for (auto i : s) {
            if (!st.empty() && st.top() == rem[0] && i == rem[1]) {
                st.pop();
                score += points;
            } else {
                st.push(i);
            }
        }
        // Collect the remaining characters
        string remaining;
        while (!st.empty()) {
            remaining.push_back(st.top());
            st.pop();
        }
        reverse(remaining.begin(), remaining.end());
        s = remaining; // Update the original string with the remaining characters
        return score;
    }
    int maximumGain(string s, int x, int y) {
        // We prioritize removing the substring with the higher points first
        int points = 0;
        if (x >= y) {
            points += removeSubstr(s, "ab", x);
            points += removeSubstr(s, "ba", y);
        } else {
            points += removeSubstr(s, "ba", y);
            points += removeSubstr(s, "ab", x);
        }
        return points;
    }
};
